### PR TITLE
[Refactor]: Add typescript

### PR DESCRIPTION
**Pull request by mcfdez**
This pull request introduces a significant shift in the codebase from JavaScript to TypeScript. It includes changes to ESLint configurations, updates to the Node.js version, changes to the GitHub Actions workflow, and updates to the Dockerfile. The JavaScript files have been rewritten in TypeScript, with the necessary changes to syntax and type annotations.

Here are the most important changes:

ESLint and Prettier Configurations:

* [`.eslintrc`](diffhunk://#diff-e7e195d3d8fa5b82fa51ed952f3a179d429cbc46c0432ce1f7b357c54c162822L2-R26): Updated the ESLint configuration to support TypeScript, including adding new plugins, parser, and parser options. Also, added new rules for indentation and maximum line length. [[1]](diffhunk://#diff-e7e195d3d8fa5b82fa51ed952f3a179d429cbc46c0432ce1f7b357c54c162822L2-R26) [[2]](diffhunk://#diff-e7e195d3d8fa5b82fa51ed952f3a179d429cbc46c0432ce1f7b357c54c162822L14-R41)
* [`.prettierrc`](diffhunk://#diff-663ade211b3a1552162de21c4031fcd16be99407aae5ceecbb491a2efc43d5d2L2-R3): Increased the print width and tab width for Prettier.

GitHub Actions and Dockerfile:

* [`.github/workflows/test.yml`](diffhunk://#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88L10-R10): Changed the GitHub Actions workflow to use the current directory instead of a specific action.
* [`Dockerfile`](diffhunk://#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L7-R7): Updated the Dockerfile ENTRYPOINT to point to the `dist` directory.

JavaScript to TypeScript:

* `src/github/write-comment.js` to `src/github/write-comment.ts`: Rewritten in TypeScript with necessary changes to syntax and type annotations. [[1]](diffhunk://#diff-b1411d2790bf60627a83cd884f791b220d3147dc741849c1c641693cab4c31b7L1-L12) [[2]](diffhunk://#diff-00b3d787461c5e814ee151e18b9239e283b3f542cbc06a36ae770d7acbb4b3d0R1-R14)
* `src/index.js` to `src/index.ts`: Rewritten in TypeScript with necessary changes to syntax and type annotations. [[1]](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L1-L27) [[2]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R1-R24)
* `src/parameters-schema.js` to `src/parameters-schema.ts`: Rewritten in TypeScript with necessary changes to syntax and type annotations. [[1]](diffhunk://#diff-eaa9407edf7c218d973dd3d9e6bd652d0c83d519a34e154f6944d2cd95ec3404L1-L15) [[2]](diffhunk://#diff-dabb2dbe6761a3e4dfd2ced85e8933b506e902d965c726731e05ef23d348e043R1-R17)
* `src/run-action.js` to `src/run-action.ts`: Rewritten in TypeScript with necessary changes to syntax and type annotations. [[1]](diffhunk://#diff-e1c6e2cd5c959fd33ba61124688691afc6546d13af4604607295c2d34b5104b4L1-L34) [[2]](diffhunk://#diff-4b7bd412f31e42e72cce0f4101aa51d3262c1583476e5b0236888b4bdfb5bbb9R1-R38)
* `src/validations/validate-parameters.js` to `src/validations/validate-parameters.ts`: Rewritten in TypeScript with necessary changes to syntax and type annotations. [[1]](diffhunk://#diff-bd4c8c1ec6b8204af27c8870f33778e9d8a9cbf694fff14cff99d594dbe1248dL1-L63) [[2]](diffhunk://#diff-13e64ceb8f9ad19b2475c6fc4ba4b1173c5f17aba6ce74c3a067cf7488aeca94R1-R61)

Package.json:

* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L2-R21): Updated the package name, added a build script, and updated the dependencies and devDependencies to support TypeScript.

### CODE REVIEW

File: .nvmrc
Comment (AlbertHernandez) [inline .nvmrc]: **question (non-blocking):** is it possible to change the node version in a separate PR?

File: .prettierrc
Comment (AlbertHernandez) [inline .prettierrc]: **nitpick (non-blocking):** can we use the previous tab width? I preffer to use 2 instead of 4, this is the usual convention I have in the projects I work

File: .prettierrc
Comment (AlbertHernandez) [inline .prettierrc]: **question (non-blocking):** do we really need to increase by 2 the width?

File: dist/github/write-comment.js
Comment (AlbertHernandez) [inline dist/github/write-comment.js]: **issue (blocking):** I think this was added by mistake, this build should be done in the Dockerfile and we should not commit those files

File: package.json
Comment (AlbertHernandez) [inline package.json]: **issue (blocking):** since now the code is in typescript, most probably we should also change this start command, to have something like:

`start: ts-node index.ts`

File: package.json
Comment (AlbertHernandez) [inline package.json:2]: **praise (non-blocking):** I like this new name :)

File: package.json
Comment (AlbertHernandez) [inline package.json:4]: **issue (blocking):** also we should change this main file to be `dist/index.js`

File: package.json
Comment (AlbertHernandez) [inline package.json:11]: **question (non-blocking):** do we need to upgrade this dependency now? I was thinking if we can do this once we finish the migration to the typescript code (the same with the dev dependencies)

File: tsconfig.json
Comment (AlbertHernandez) [inline tsconfig.json]: **suggestion (non-blocking):** maybe we should point to a more recent target like ES2022

File: src/validations/validate-parameters.ts
Comment (AlbertHernandez) [inline src/validations/validate-parameters.ts]: **suggestion (non-blocking):** what do you think about creating a custom type for the schema? In this way instead of repeating this Record<any, any> we could point to the type:

`export type ValidationSchema = Record<any, any>`

Also, if its possible, would be nice to try to avoid those any and have something like:

```ts
export interface ValidationSchema {
  [key: string]: {
    required: boolean;
    availableValues?: string[]
  }
}
```

File: package.json
Comment (AlbertHernandez) [inline package.json:7]: **suggestion (if-minor):** maybe instead of tsc we can make use of SWC or we can change it in a future iteration, SWC is another compiler for typescript that is so much faster than TSC: https://swc.rs/docs/getting-started

To use it we can:

1. Install swc: `npm i --save-dev @swc/cli @swc/core`
2. Use it in the script: `swc ./src -d dist --strip-leading-paths`

File: Dockerfile
Comment (AlbertHernandez) [inline Dockerfile:8]: **issue (blocking):** dont forget to build the javascript code from the typescript one, we should add before this entrypoint: `RUN npm run build`

**Review by AlbertHernandez — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml
Comment (AlbertHernandez) [inline .github/workflows/test.yml]: **issue (blocking):** if we are going to referenciate in this way to our current action, I think we should have a previoys checkout step: https://github.com/AlbertHernandez/branch-name-action/actions/runs/7876094680/job/21495012361

File: .nvmrc
Comment (mcfdez) [reply .nvmrc]: Here is the PR with the Node version change: https://github.com/AlbertHernandez/branch-name-action/pull/27

File: .prettierrc
Comment (mcfdez) [reply .prettierrc]: I put it back as it was!

File: dist/github/write-comment.js
Comment (mcfdez) [reply dist/github/write-comment.js]: You are right. I have removed the compiled and modified the Dockerfile to compile the library.

File: package.json
Comment (mcfdez) [reply package.json]: The Dockerfile will take care of compiling the library so the command remains the same. However we can take different ways. Use ts-node, continue compiling the library, use Deno, ...

File: package.json
Comment (mcfdez) [reply package.json:11]: I have increased the versions of the dependencies in this PR since I have upgraded the Node version. If we merge the PR https://github.com/AlbertHernandez/branch-name-action/pull/27 before, I think we can keep the package this way.

File: tsconfig.json
Comment (mcfdez) [reply tsconfig.json]: Changed.

File: src/validations/validate-parameters.ts
Comment (mcfdez) [reply src/validations/validate-parameters.ts]: Totally agree in strongly typing the application. I had not done it in this PR as I was planning to do it in successive PRs. But since it is necessary to take advantage of all the potential that Typescript offers us, I have included this typing.

File: package.json
Comment (mcfdez) [reply package.json:7]: I had thought of adding SWC or ESBuild in a second iteration.

File: Dockerfile
Comment (mcfdez) [reply Dockerfile:8]: Added

File: .github/workflows/test.yml
Comment (mcfdez) [reply .github/workflows/test.yml]: Right. Added

File: package.json
Comment (AlbertHernandez) [reply package.json:7]: SWC in a second iteration sounds great, thanks!

File: src/validations/validate-parameters.ts
Comment (AlbertHernandez) [reply src/validations/validate-parameters.ts]: Awesome, thanks!

File: .nvmrc
Comment (AlbertHernandez) [reply .nvmrc]: Nice, then let's remove this change from this PR, since we already have the other for changing the node version

File: package.json
Comment (AlbertHernandez) [reply package.json]: Yes, for now let's use `ts-node index.ts` that most probably is the simpler solution, in a future iteration we can see other alternatives.

I am suggesting this change because if someone tries to run `npm run start` with the current solution, it will get an error indicating this files does not exist.

File: package.json
Comment (AlbertHernandez) [reply package.json:11]: If we need to change the version of the library due to node version, most probably this change should be done in the other PR and not here

File: src/run-action.ts
Comment (AlbertHernandez) [inline src/run-action.ts:13]: **thought (non-blocking):** I didn't know it can happen the branch name to be optional, if this is the case good catch

File: src/validations/validate-parameters.ts
Comment (AlbertHernandez) [inline src/validations/validate-parameters.ts]: **suggestion (non-blocking):** We should also update this type to use `ValidationSchema`

File: src/validations/validate-parameters.ts
Comment (AlbertHernandez) [inline src/validations/validate-parameters.ts]: **suggestion (if-minor):** maybe I would move this interface to a separate file to keep the files simple and reduce the logic we have, because if we have it in this current way we are going to end up having the logic about how to validate and the logic about how to define a schema in the same file, which I am not sure is the best. Anyway, this is a minor change we can always do later.

File: .github/workflows/test.yml
Comment (AlbertHernandez) [inline .github/workflows/test.yml:13]: **issue (blocking):** I see this still failing (https://github.com/AlbertHernandez/branch-name-action/actions/runs/7882599800/job/21532232107), not sure if we should use the ref of the current context or other thing, let's investigate how is the best way to point to our current action, if not, we can always have this in the previous way and improve it in the next PR

File: package.json
Comment (mcfdez) [reply package.json]: Changed.

File: package.json
Comment (mcfdez) [reply package.json:11]: Changed in the other PR the dependency versions.

File: src/validations/validate-parameters.ts
Comment (mcfdez) [reply src/validations/validate-parameters.ts]: Changed

File: src/validations/validate-parameters.ts
Comment (mcfdez) [reply src/validations/validate-parameters.ts]: Extracted to another file

File: .github/workflows/test.yml
Comment (mcfdez) [reply .github/workflows/test.yml:13]: This error appears because you are trying to launch an action whose tag does not exist in the main branch. With the change I provide, the test workflow will trigger the local action. However, it cannot be tested until the change reaches the main branch.

Comment (AlbertHernandez) [reply .github/workflows/test.yml:13]: I think we can solve this issue pointing to the current branch, this checkout action also accept a property named ref, if this ref is the branch name we are working, most probably this will work

File: package.json
Comment (AlbertHernandez) [reply package.json:11]: Nice, then we will wait until we have the other PR merged to continue with this one

File: src/models/validation-schema.model.ts
Comment (AlbertHernandez) [inline src/models/validation-schema.model.ts:2]: **question (non-blocking):** is this formatted properly? I see indentation of 4 spaces instead of 2

File: src/validations/validate-parameters.ts
Comment (AlbertHernandez) [reply src/validations/validate-parameters.ts]: Nice thanks!

**Pull request closed without merging**
